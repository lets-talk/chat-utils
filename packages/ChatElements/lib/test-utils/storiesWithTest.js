import { withTests } from '@storybook/addon-jest';
import jestTestResults from '../../tests-results/jest-test-results.json';

export default withTests(jestTestResults, {
  filesExt: '__tests__/**.js',
});
