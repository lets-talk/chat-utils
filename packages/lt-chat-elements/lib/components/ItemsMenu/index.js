import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { rgba } from 'polished';

import { textColor, themeColor } from '../../utils/style';

const StyledItemsList = styled.ul`
  display: block;
  cursor: pointer;
  padding: 0;
  margin: 0;
  border-radius: 5px;
  background-color: ${(props) => props.theme.palette.common.white};
  font-size: ${(props) => props.theme.typography.classes.caption.fontSize};
  line-height: ${(props) => props.theme.typography.classes.caption.lineHeight};
  font-weight: ${(props) => props.theme.typography.weights.fontWeightMedium};
  box-shadow: ${(props) => props.theme.shadows['4p']};
`;

const StyledListItem = styled.li`
  list-style: none;
  padding: 15px;
  color: ${(props) => textColor(props.theme, 'light', 'secondary')};

  &:hover {
    color: ${(props) => themeColor(props.theme, 'accent', 'base')};
    background-color: ${(props) => rgba(themeColor(props.theme, 'accent', 'base'), 0.1)};
  }

  ${(props) =>
    props.withDivider &&
      css`
        &:not(:first-child) {
          border-top: 1px solid #e0e0e0;
        }
      `};
`;
StyledListItem.displayName = 'StyledListItem';

const ItemsMenu = (props) => {
  const {
    items, onItemClick, render, withDivider,
  } = props;

  return (
    <StyledItemsList>
      {items.map((item, index) =>
        (
          <StyledListItem
            key={item.id}
            role="button"
            tabIndex={index}
            withDivider={withDivider}
            onClick={onItemClick && ((e) => onItemClick(item, index, e))}
            onKeyUp={onItemClick && ((e) => onItemClick(item, index, e))}
          >
            {render(item)}
          </StyledListItem>
        ))}
    </StyledItemsList>
  );
};

ItemsMenu.defaultProps = {
  onItemClick: null,
  className: 'LT-ItemsMenu-Container',
};

ItemsMenu.propTypes = {
  /**
   * Render prop is called passing each item
   */
  render: PropTypes.func.isRequired,
  /**
   * Array of elements to display
   */
  items: PropTypes.array.isRequired,
  /**
   * Whether to show or not an horizontal line dividing each item
   */
  withDivider: PropTypes.bool,
  /**
   * Extra className to style the component
   */
  /* eslint-disable react/no-unused-prop-types */
  className: PropTypes.string,
  /**
   * Callback function called when user clicks on an action Item
   * The callback has the following signature:
   * callback (action, index, e)
   */
  onItemClick: PropTypes.func,
};

ItemsMenu.defaultProps = {
  className: 'LT-ItemsMenu-Wrapper',
  withDivider: true,
};

ItemsMenu.displayName = 'ItemsMenu';

export default ItemsMenu;
