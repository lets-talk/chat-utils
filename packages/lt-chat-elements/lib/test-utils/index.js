// test-utils/index.js

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import { ThemeProvider } from 'styled-components';

const defaultTheme = {
  palette: {
    common: {
      black: 'rgb(0, 0, 0)',
      white: 'rgb(255, 255, 255)',
      online: 'rgb(67, 160, 71)',
      offline: 'rgb(229, 57, 53)',
      sleeping: 'rgb(253, 216, 53)',
    },
    colors: {
      primary: {
        base: 'rgb(4, 38, 73)',
        light: 'rgb(86, 165, 246)',
        dark: 'rgb(0, 0, 0)',
      },
      accent: {
        base: 'rgb(255, 87, 50)',
        light: 'rgb(255, 171, 152)',
        dark: 'rgb(203, 37, 0)',
      },
      foreground: {
        base: 'rgb(94, 124, 139)',
        light: 'rgb(151, 173, 184)',
        dark: 'rgb(53, 70, 78)',
      },
      background: {
        base: 'rgb(255, 255, 255)',
        light: 'rgb(255, 255, 255)',
        dark: 'rgb(204, 204, 204)',
      },
    },
    contrastColors: {
      dark: {
        primary: 'rgb(255, 255, 255)',
        secondary: 'rgba(255, 255, 255, 0.7)',
        disabled: 'rgba(255, 255, 255, 0.5)',
        hint: 'rgba(255, 255, 255, 0.12)',
      },
      light: {
        primary: 'rgba(0, 0, 0, 0.87)',
        secondary: 'rgba(0, 0, 0, 0.54)',
        disabled: 'rgba(0, 0, 0, 0.38)',
        hint: 'rgba(0, 0, 0, 0.12)',
      },
    },
    error: {
      light: '#e57373',
      main: '#f44336',
      dark: '#d32f2f',
      contrastText: '#fff',
    },
    divider: 'rgba(255, 255, 255, 0.1)',
    action: {
      active: 'rgba(255, 255, 255, 0.9)',
      hover: 'rgba(255, 255, 255, 0.08)',
      selected: 'rgba(255, 255, 255, 0.14)',
      disabled: 'rgba(255, 255, 255, 0.26)',
      disabledBackground: 'rgba(255, 255, 255, 0.12)',
    },
  },
  shadows: {
    '2p': '0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 3px 1px -2px rgba(0, 0, 0, 0.2), 0 1px 5px 0 rgba(0, 0, 0, 0.12)',
    '3p': '0 3px 4px 0 rgba(0, 0, 0, 0.14), 0 3px 3px -2px rgba(0, 0, 0, 0.2), 0 1px 8px 0 rgba(0, 0, 0, 0.12)',
    '4p': '0 4px 5px 0 rgba(0, 0, 0, 0.14), 0 1px 10px 0 rgba(0, 0, 0, 0.12), 0 2px 4px -1px rgba(0, 0, 0, 0.2)',
    '6p': '0 6px 10px 0 rgba(0, 0, 0, 0.14), 0 1px 18px 0 rgba(0, 0, 0, 0.12), 0 3px 5px -1px rgba(0, 0, 0, 0.2)',
    '8p': '0 8px 10px 1px rgba(0, 0, 0, 0.14), 0 3px 14px 2px rgba(0, 0, 0, 0.12), 0 5px 5px -3px rgba(0, 0, 0, 0.2)',
    '16p': '0 16px 24px 2px rgba(0, 0, 0, 0.14), 0 6px 30px 5px rgba(0, 0, 0, 0.12), 0 8px 10px -5px rgba(0, 0, 0, 0.2)',
  },
  typography: {
    fontFamily: "'Nimbus sans-serif', 'Sans', 'Helvetica', 'Arial'",
    fontSize: 'calc(1 * 1px)',
    weights: {
      fontWeightUltraThin: 100,
      fontWeightThin: 300,
      fontWeightNormal: 300,
      fontWeightMedium: 500,
      fontWeightSemibold: 600,
      fontWeightBold: 700,
    },
    classes: {
      body: {
        fontSize: 'calc(16 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
      caption: {
        fontSize: 'calc(12 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
      footnote: {
        fontSize: 'calc(14 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
      small: {
        fontSize: 'calc(10 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
      headline: {
        fontSize: 'calc(24 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
      subhead: {
        fontSize: 'calc(18 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
      title: {
        fontSize: 'calc(20 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
      jumbo: {
        fontSize: 'calc(44 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
      mega: {
        fontSize: 'calc(56 * 1px)',
        lineHeight: 'calc(16 * 1px)',
      },
    },
  },
  components: {
    conversationBox: {
      externalBackgroundColor: 'rgba(94, 124, 139, 0.1)',
      internalBackgroundColor: 'rgba(255, 213, 0, 0.2)',
      importantBackgroundColor: 'rgb(255, 246, 244)',
    },
    loader: {
      backgroundColor: 'rgb(252, 88, 48)',
    },
  },
};

const shallowWithTheme = (component) => {
  const context = shallow(<ThemeProvider theme={defaultTheme} />)
    .instance()
    .getChildContext();
  return shallow(component, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};

const mountWithTheme = (component) => {
  const context = shallow(<ThemeProvider theme={defaultTheme} />)
    .instance()
    .getChildContext();

  return mount(component, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};

const renderWithTheme = (component) => renderer.create(<ThemeProvider theme={defaultTheme}>{component}</ThemeProvider>);

const staticWithTheme = (component) => render(<ThemeProvider theme={defaultTheme}>{component}</ThemeProvider>);

export {
  defaultTheme,
  shallowWithTheme,
  mountWithTheme,
  staticWithTheme,
  renderWithTheme,
};
