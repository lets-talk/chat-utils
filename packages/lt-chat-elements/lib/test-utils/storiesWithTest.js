import { withTests } from '@storybook/addon-jest';
import jestTestResults from '../../tests-results/jest-test-results.json';

export default withTests({ results: jestTestResults }, {
  filesExt: '__tests__/**.js',
});
