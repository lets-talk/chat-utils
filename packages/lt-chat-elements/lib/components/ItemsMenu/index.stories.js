import React from 'react';
// Storybook stuff
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, select, object } from '@storybook/addon-knobs';

// Component to show on storybook
import ItemsMenu from '.';
import { Avatar } from '../../../lib';
import { WrapWithTheme, defaultTheme, themeOptions } from '../../utils/stories';
import withTests from '../../test-utils/storiesWithTest';

const stories = storiesOf('ItemsMenu', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests('ItemsMenu'));

const items = [
  {
    id: 1,
    name: 'Parent 1',
    greeting: '',
  },
  {
    id: 2,
    name: 'Parent 2',
    greeting: '',
  },
  {
    id: 3,
    name: 'Child 1.1',
    greeting: '',
  },
  {
    id: 4,
    name: 'Child 1.2',
    greeting: '',
  },
  {
    id: 5,
    name: 'Child 1.2.1',
    greeting: '',
  },
];

const render = (element) => {
  const { name } = element;
  return (
    <span>
      <Avatar src="http://i46.tinypic.com/sexbb8.png" size="xsmall" />{name}
    </span>
  );
};

stories.addWithInfo(
  'ItemsMenu',
  'This is the basic usage with the ItemsMenu',
  () => (
    <div style={{ position: 'fixed', bottom: 0, right: '10px' }}>
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <ItemsMenu
          withDivider={boolean('withDivider', true)}
          onItemClick={action('onItemClick')}
          items={object('Items', items)}
          render={render}
        />
      </WrapWithTheme>
    </div>
  )
);
