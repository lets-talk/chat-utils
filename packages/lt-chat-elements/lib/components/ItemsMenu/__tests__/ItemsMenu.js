import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

import { renderWithTheme, staticWithTheme, shallowWithTheme } from '../../../test-utils';
import ItemsMenu from '../index';

describe('ItemsMenu component', () => {
  const emptyData = [];

  const items = [
    {
      id: 1,
      name: 'Parent 1',
      greeting: '',
      hidden: false,
      internal: false,
      parent_id: null,
      priority: 1,
    },
    {
      id: 2,
      name: 'Parent 2',
      greeting: '',
      hidden: false,
      internal: false,
      parent_id: null,
      priority: 1,
    },
    {
      id: 3,
      name: 'Child 1.1',
      greeting: '',
      hidden: false,
      internal: false,
      parent_id: 1,
      priority: 1,
    },
    {
      id: 4,
      name: 'Child 1.2',
      greeting: '',
      hidden: false,
      internal: false,
      parent_id: 1,
      priority: 1,
    },
    {
      id: 5,
      name: 'Child 1.2.1',
      greeting: '',
      hidden: false,
      internal: false,
      parent_id: 4,
      priority: 1,
    },
  ];

  const render = (element) => {
    const { name } = element;
    return (
      <div>{name}</div>
    );
  };

  it('should render with empty data', () => {
    const component = shallow(<ItemsMenu items={emptyData} render={render} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with data', () => {
    const component = shallow(<ItemsMenu items={items} render={render} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should have the correct styles (Those that depends on the theme)', () => {
    const tree = renderWithTheme(<ItemsMenu items={items} render={render} />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('background-color', 'rgb(255,255,255)');
    expect(tree).toHaveStyleRule('font-size', 'calc(12 * 1px)');
    expect(tree).toHaveStyleRule('line-height', 'calc(16 * 1px)');
    expect(tree).toHaveStyleRule('font-weight', '500');
  });

  it('should render the correct html structure', () => {
    const wrapper = staticWithTheme(<ItemsMenu items={items} render={render} />);
    expect(wrapper.find('li')).toHaveLength(5);
  });

  it('should handle the click event on the items correctly', () => {
    const onItemClick = jest.fn();
    const fakeEventObject = { preventDefault() {}, stopPropagation() {} };
    // const wrapper = mountWithTheme(<ItemsMenu items={items} onItemClick={onItemClick} />);
    const wrapper = shallowWithTheme(<ItemsMenu items={items} render={render} onItemClick={onItemClick} />);
    wrapper.find('StyledListItem').last().simulate('click');
    // The function is called with the item clicked, the index and the event (item, index, e)
    expect(onItemClick).toHaveBeenCalledWith(items[4], 4, undefined);

    onItemClick.mockClear();
    wrapper.find('StyledListItem').first().simulate('click');
    // The function is called with the item clicked, the index and the event (item, index, e)
    expect(onItemClick).toHaveBeenCalledWith(items[0], 0, undefined);

    onItemClick.mockClear();
    wrapper.find('StyledListItem').at(1).simulate('click', fakeEventObject);
    // The function is called with the item clicked, the index and the event (item, index, e)
    expect(onItemClick).toHaveBeenCalledWith(items[1], 1, fakeEventObject);
  });
});
