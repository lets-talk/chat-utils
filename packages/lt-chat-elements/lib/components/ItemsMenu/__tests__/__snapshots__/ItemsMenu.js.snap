// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsMenu component should have the correct styles (Those that depends on the theme) 1`] = `
.c0 {
  display: block;
  cursor: pointer;
  padding: 0;
  margin: 0;
  border-radius: 5px;
  background-color: rgb(255,255,255);
  font-size: calc(12 * 1px);
  line-height: calc(16 * 1px);
  font-weight: 500;
  box-shadow: 0 4px 5px 0 rgba(0,0,0,0.14),0 1px 10px 0 rgba(0,0,0,0.12),0 2px 4px -1px rgba(0,0,0,0.2);
}

.c0 li {
  list-style: none;
  padding: 15px;
  color: rgba(0,0,0,0.54);
}

.c0 li:hover {
  color: rgb(255,87,50);
  background-color: rgba(255,87,50,0.1);
}

.c1 {
  margin: 0;
  margin-top: -1px;
  margin-left: 0;
  height: 1px;
  border: none;
  background-color: #e0e0e0;
}

<ul
  className="c0"
>
  <div
    className=""
    onClick={undefined}
    onKeyUp={undefined}
    role="button"
    tabIndex={0}
  >
    <li>
      <div>
        Parent 1
      </div>
    </li>
    <hr
      className="c1"
    />
  </div>
  <div
    className=""
    onClick={undefined}
    onKeyUp={undefined}
    role="button"
    tabIndex={1}
  >
    <li>
      <div>
        Parent 2
      </div>
    </li>
    <hr
      className="c1"
    />
  </div>
  <div
    className=""
    onClick={undefined}
    onKeyUp={undefined}
    role="button"
    tabIndex={2}
  >
    <li>
      <div>
        Child 1.1
      </div>
    </li>
    <hr
      className="c1"
    />
  </div>
  <div
    className=""
    onClick={undefined}
    onKeyUp={undefined}
    role="button"
    tabIndex={3}
  >
    <li>
      <div>
        Child 1.2
      </div>
    </li>
    <hr
      className="c1"
    />
  </div>
  <div
    className=""
    onClick={undefined}
    onKeyUp={undefined}
    role="button"
    tabIndex={4}
  >
    <li>
      <div>
        Child 1.2.1
      </div>
    </li>
    <hr
      className="c1"
    />
  </div>
</ul>
`;

exports[`ItemsMenu component should render with data 1`] = `
<styled.ul>
  <StyledListItem
    role="button"
    tabIndex={0}
  >
    <li
      key="1"
    >
      <div>
        Parent 1
      </div>
    </li>
    <styled.hr />
  </StyledListItem>
  <StyledListItem
    role="button"
    tabIndex={1}
  >
    <li
      key="2"
    >
      <div>
        Parent 2
      </div>
    </li>
    <styled.hr />
  </StyledListItem>
  <StyledListItem
    role="button"
    tabIndex={2}
  >
    <li
      key="3"
    >
      <div>
        Child 1.1
      </div>
    </li>
    <styled.hr />
  </StyledListItem>
  <StyledListItem
    role="button"
    tabIndex={3}
  >
    <li
      key="4"
    >
      <div>
        Child 1.2
      </div>
    </li>
    <styled.hr />
  </StyledListItem>
  <StyledListItem
    role="button"
    tabIndex={4}
  >
    <li
      key="5"
    >
      <div>
        Child 1.2.1
      </div>
    </li>
    <styled.hr />
  </StyledListItem>
</styled.ul>
`;

exports[`ItemsMenu component should render with empty data 1`] = `<styled.ul />`;
